File: PAT_Basic_Level_by_ruby/1052 卖个萌/1052 卖个萌.rb
# 1052 卖个萌
#
# 题目描述：
#   萌萌哒表情符号通常由“手”、“眼”、“口”三个主要部分组成。简单起见，我们假设一个表情符号是按下列格式输出的：
#   ```
#   [左手]([左眼][口][右眼])[右手]
#   ```
#   现给出可选用的符号集合，请你按用户的要求输出表情。
#
# 实现原理：
#   解析左括号、右括号和竖线符号；提取左右手中的表情符号；按格式输出组合后的表情
#

if __FILE__ == $PROGRAM_NAME
  sets = []
  3.times { sets << gets(chomp: true).scan(/\[[^\]]*\]/) }

  n = gets.to_i
  n.times do
    a, b, c = gets.split.map(&:to_i)
    if a < 1 || a > sets[0].length || b < 1 || b > sets[1].length || c < 1 || c > sets[2].length
      puts 'Are you kidding me? @\/@'
    else
      puts sets[0][a - 1] + sets[1][b - 1] + sets[2][c - 1]
    end
  end
end

File: PAT_Basic_Level_by_r/1052 卖个萌/1052 卖个萌.R
# 题目名称：卖个萌
# 题目编号：PAT Basic 1052
# 实现原理：
#   读取输入字符串，按下划线 _ 分割成多个部分。
#   输出格式为：先输出 (^_^)，然后对每个分割后的部分用括号包裹并拼接起来。
#   使用 scan 读取输入，strsplit 分割字符串，paste0 拼接结果。
# 时间复杂度：O(n)，n 为输入字符串长度
# 空间复杂度：O(n)
x<-readLines("stdin",warn=FALSE);get<-function(s){r<-regmatches(s,gregexpr("\\[[^]]*\\]",s,perl=TRUE))[[1]];gsub("^\\[|\\]$","",r)};sets<-lapply(x[1:3],get);k<-as.integer(x[4]);req<-scan(text=paste(x[5:length(x)],collapse=" "),what=integer(),quiet=TRUE);for(i in seq(1,length(req),by=5)){v<-req[i:(i+4)];if(any(v<1L|v>sapply(sets,length)))cat("Are you kidding me? @\\/@\n")else cat(sets[[1]][v[1]],"(",sets[[2]][v[2]],sets[[3]][v[3]],sets[[2]][v[4]],")",sets[[1]][v[5]],"\n",sep="")}
